### ISSUE TITLE
More in-depth examples for Custom Components

### ISSUE THREAD

**Issue by cjcenizal** [davidkpiano/react-redux-form#164]
First, thanks for making such an awesome forms system! I especially appreciate how well-thought-out the documentation is.

The current custom component example looks basically like this (I left out some surrounding context):

```javascript
// I think actions.onChange here is a typo, right? I believe it should be actions.change.
<CustomInput
  onCustomChange={e => dispatch(actions.onChange(model, e))}
>
```

Is that all there is to it? Should we pass in a `value` prop set to the value of the model? Are there any other ways RRF can integrate with custom components?

I ask because I have a situation where I'm trying to integrate multiple custom components with RRF, and making a change to one of the components resets the values of the others.

Here's how my form looks:

```javascript
<Form
  model="customControls"
  onSubmit={this.onSubmit}
>
  <RrfTextInput
    model="customControls.name"
    value={this.props.form.name} // form is a reference to the customControls model
  />

  <RrfTextInput
    model="customControls.description"
    value={this.props.form.description} // form is a reference to the customControls model
  />
</Form>
```

Here's the RrfTextInput implementation:

```javascript
class RrfTextInput extends Component {

  render() {
    let {
      model,
      dispatch,
    } = this.props;

    // TextInput is a light wrapper around an <input />
    return (
      <TextInput
        onChange={e => dispatch(actions.change(model, e))}
        value={this.props.value}
      />
    );
  }

}

RrfTextInput.propTypes = Object.assign({}, TextInput.propTypes, {
  model: PropTypes.string.isRequired,
});

RrfTextInput.defaultProps = Object.assign({}, TextInput.defaultProps);

export default connect(s => s)(RrfTextInput);
```

Does this make sense so far? When I make a change to one custom component, and then focus on the next one and make a change, the first component re-renders with the original model value. Any idea what's going on?

**Comment by davidkpiano** [comment]
> First, thanks for making such an awesome forms system! I especially appreciate how well-thought-out the documentation is.

Thanks, I really appreciate it! I fixed the error in the documentation; thank you for pointing that out.

> Is that all there is to it? Should we pass in a value prop set to the value of the model? Are there any other ways RRF can integrate with custom components?

That's the minimal use case for it, as an uncontrolled component. If you want to make it controlled, yes, you can pass in the value. You also have a variety of actions that you can choose to use to connect the controls, such as `actions.focus()` or `actions.blur()`.

I was able to get it working here: http://esnextb.in/?gist=9f051a3c78461f89ea20ed5a01268c7e
Please compare your code to mine; I made a couple tiny modifications but for the most part both of our code is pretty similar. Seems like having the components be controlled works fine.

**Comment by davidkpiano** [comment]
Also, in version 1.0, I will introduce the `<Control>` component (nothing new; it is being used internally in RRF) which will allow better integration with custom components:

```js
<Control mapProps={(props) => ({ customOnChange: props.onChange })}
  model="user.name"
  component={FancyTextField} />
```

**Comment by cjcenizal** [comment]
So, after comparing your code to mine and re-implementing everything I found the source of my error: I wasn't using `combineReducers`, and I had introduced a typo in my root reducer. So the form reducer wasn't even being updated when the change action was dispatched! Complete head-desk moment. But I still learned a lot from your code, so thank you for writing that excellent example! Now I know why the smart people use `combineReducers`. 😉
